flink-core/src/main/java/org/apache/flink/core/fs/LimitedConnectionsFileSystem.java

```diff
@@ -0,0 +1,1097 @@
+			deadline = deadlineNanos > 0 ? deadlineNanos : Long.MAX_VALUE;
+		}
+
+		// wait for available connections
+		long timeLeft;
+
+		if (streamInactivityTimeoutNanos == 0) {
+			// simple case: just wait
+			while ((timeLeft = (deadline - System.nanoTime())) > 0 &&
+					!hasAvailability(totalLimit, outputLimit, inputLimit)) {
+
+				available.await(timeLeft, TimeUnit.NANOSECONDS);
+			}
+		}
+		else {
+			// complex case: chase down inactive streams
+			final long checkIntervalNanos = (streamInactivityTimeoutNanos >>> 1) + 1;
+
+			long now;
+			while ((timeLeft = (deadline - (now = System.nanoTime()))) > 0 && // while still within timeout
+					!hasAvailability(totalLimit, outputLimit, inputLimit)) {
+
+				// check all streams whether there in one that has been inactive for too long
+				if (!(closeInactiveStream(openOutputStreams, now) || closeInactiveStream(openInputStreams, now))) {
+					// only wait if we did not manage to close any stream.
+					// otherwise eagerly check again if we have availability now (we should have!)
+					long timeToWait = Math.min(checkIntervalNanos, timeLeft);
+					available.await(timeToWait, TimeUnit.NANOSECONDS);
+				}
+			}
+		}
+
+		// check for timeout
+		// we check availability again to catch cases where the timeout expired while waiting
+		// to re-acquire the lock
+		if (timeLeft <= 0 && !hasAvailability(totalLimit, outputLimit, inputLimit)) {
+			throw new IOException(String.format(
+					"Timeout while waiting for an available stream/connect. " +
+					"limits: total=%d, input=%d, output=%d ; Open: input=%d, output=%d ; timeout: %d ms",
+					maxNumOpenStreamsTotal, maxNumOpenInputStreams, maxNumOpenOutputStreams,
+					numReservedInputStreams, numReservedOutputStreams, getStreamOpenTimeout()));
+		}
+	}
+
+	@GuardedBy("lock")
+	private boolean hasAvailability(int totalLimit, int outputLimit, int inputLimit) {
+		return numReservedOutputStreams < outputLimit &&
+				numReservedInputStreams < inputLimit &&
+				numReservedOutputStreams + numReservedInputStreams < totalLimit;
+	}
+
+	@GuardedBy("lock")
+	private boolean closeInactiveStream(HashSet<? extends StreamWithTimeout> streams, long nowNanos) {
+		for (StreamWithTimeout stream : streams) {
+			try {
+				// If the stream is closed already, it will be removed anyways, so we
+				// do not classify it as inactive. We also skip the check if another check happened too recently.
+				if (stream.isClosed() || nowNanos < stream.getLastCheckTimestampNanos() + streamInactivityTimeoutNanos) {
+					// interval since last check not yet over
+					return false;
+				}
+				else if (!stream.checkNewBytesAndMark(nowNanos)) {
+					stream.closeDueToTimeout();
+					return true;
+				}
+			}
+			catch (StreamTimeoutException ignored) {
+				// may happen due to races
+			}
+			catch (IOException e) {
+				// only log on debug level here, to avoid log spamming
+				LOG.debug("Could not check for stream progress to determine inactivity", e);
```

Comment:
I wonder what to do with a stream that throws exceptions on `getPos` and avoids check for inactivity but also is not closed. This looks like a candidate that is in a bad state. While it will probably sooner or later also cause exceptions somewhere else, I think it could theoretically dodge closing forever and block attempts to create new streams. Might be appropriate to close such a stream proactively?

Reply:
Good point, I thought about that as well, but was not sure. I would feel a bit hesitant to do that on the first time such an exception occurs, and it would need more elaborate logic otherwise.